demos/GMW_demo/gmw_and.js

```diff
@@ -0,0 +1,175 @@
+  }
+
+  // AND the two shares when ready
+  var ready = function () {
+    var final_deferred = new jiff.helpers.Deferred();
+    var final_promise = final_deferred.promise;
+    var wi= share1.value & share2.value;
+    var csecret=[];
+    csecret.push(share1.value);
+    csecret.push(share2.value);
+    // receivinglist of OT msg
+    var recls=[];
+    var i;
+    for ( i=1;i<=jiff.party_count;i++) {
+      if (i<jiff.id) {
+        recls.push(i);
+      }
+    }
+
+    var share_id ;
+    if (share_id==null) {
+      share_id= jiff.counters.gen_op_id2('otshare', share1.holders, share1.holders);
+    }
+    var my_count;//list of promises needs to be resolve
+    if (typeof (my_count) === 'undefined') {
+      my_count= initial_deferreds(jiff,share_id);
+    }
+    // generate OT msg
+    var four_opts={};
+    four_opts['opts']=OTGate(csecret);// jason object
+    four_opts['sender_id']=jiff.id;
+    four_opts['op_id']=share_id;
+    four_opts = jiff.hooks.execute_array_hooks('beforeOperation', [jiff, 'open', four_opts], 2);
+    var mymsg=JSON.stringify(four_opts);
+    var help=[];
+    help.push(wi);
+    //wi resolve local wi; special case: jiff.id===jiff.party_count, return value
+    jiff.deferreds[share_id][jiff.id].resolve(help);
+
+    // OT send process, send OT msg to small party_ids
+    jiff.emit('OT',recls,mymsg,true);
+
+    // OT receive process receive from larger party_ids' OT msg
+    for ( var ssid=jiff.id+1;ssid<=jiff.party_count;ssid++) {
+
+      jiff.listen('OT',function (ssid,msg) {
+        msg=JSON.parse(msg);
+        var sendId=msg['sender_id'];
+        var opId=msg['op_id'];
+        var opts=msg['opts'];
+        if (jiff.deferreds[opId] == null) {
+          jiff.deferreds[opId] = {};
+        }
+        if (jiff.deferreds[opId][sendId] == null) { // not ready, setup a deferred
+          jiff.deferreds[opId][sendId] = new jiff.helpers.Deferred();
+        }
+        jiff.deferreds[opId][sendId].resolve(opts);
+      });
+    }
+    // compute ci to return
+    Promise.all(my_count).then(function (v) {
+      //v= [ [ 0 ], [ 0, 0, 0, 0 ], [ 0, 1, 0, 1 ] ]
+
+      // choose one result from 4 options
+      var find = function (msgls,csecret) {
+        var choose=2*csecret[0]+csecret[1];// 0,1,2,3
+        var one=msgls[choose];
+        return one;
+      }
+
+      var re=0;
+      for ( i=0;i<v.length;i++) {
+        var cur=v[i];
+        if (cur.length===1) {
+          re=re^cur[0];
+        } else {
+          var ans=find(cur,csecret);// ot select one answer from 4 options
+          re=re^ans;
+        }
+      }
+
+      var _remaining=jiff.party_count-jiff.id+1;
+      for (i =jiff.id;i<=jiff.party_count;i++) {
+        if (jiff.deferreds[share_id][i]!=='undefined') {
+          var mpromise=jiff.deferreds[share_id][i].promise;
+          (function (mpromise, i) { // p_id is modified in a for loop, must do this to avoid scoping issues.
```

Comment:
Try and remove references to `jiff.deferreds` within the `GMW_and` function itself, letting that logic be handled by `listen`, `OT.receive` / `get`, etc. Because above you awaited the array of results from OT.receive, this part of the function should be able do the local computations (i.e. lines 90-98) and then resolve the deferred (i.e line 116).